File: leetcode_problems/cpp/lc744_find_smallest_letter_larger_than_target.cpp
#include <vector>
using std::vector;

class Solution
{
public:
    bool is_good(char value, char target)
    {
        return (value - 'a') <= (target - 'a');
    }

    char nextGreatestLetter(vector<char> &letters, char target)
    {
        int low = -1;
        int high = letters.size();
        while (high - low > 1)
        {
            int mid = low + (high - low) / 2;
            if (is_good(letters[mid], target))
            {
                low = mid;
            }
            else
            {
                high = mid;
            }
        }
        return high == letters.size() ? letters[0] : letters[high];
    }
};

File: leetcode_problems/typescript/src/lc744_find_smallest_letter_larger_than_target.ts
function nextGreatestLetterIsGood(value: string, target: string): boolean {
  return (
    value.charCodeAt(0) - "a".charCodeAt(0) <=
    value.charCodeAt(0) - "a".charCodeAt(0)
  );
}

function nextGreatestLetter(letters: string[], target: string): string {
  let low = -1;
  let high = letters.length;
  while (high - low > 1) {
    let mid = low + Math.floor((high - low) / 2);
    if (nextGreatestLetterIsGood(letters[mid], target)) {
      low = mid;
    } else {
      high = mid;
    }
  }
  return high === letters.length ? letters[0] : letters[high];
}
